src/builders/thermodynamic.jl

```diff
@@ -0,0 +1,258 @@
+                bound = log(ratio),
+            )
+    end
+
+    m
+end
+
+export add_metabolite_ratio_constraints!
+
+"""
+$(TYPEDSIGNATURES)
+
+Perform a max-min driving force analysis on the `model`, as defined by Noor, et al.,
+"Pathway thermodynamics highlights kinetic obstacles in central metabolism.", PLoS
+computational biology, 2014.
+
+The function uses the supplied `optimizer` and `reaction_standard_gibbs_free_energies`.
+Optionally, `flux_solution` can be used to set the directions of each reaction in `model`
+(all reactions are assumed to proceed forward and are active by default). The supplied
+`flux_solution` should be free of internal cycles i.e. thermodynamically consistent. This
+optional input is important if a reaction in `model` normally runs in reverse (negative
+flux). Note, reactions in `flux_solution` that are smaller than `small_flux_tol` are also
+ignored in the analysis function (for numerical stability).
+
+The max-min driving force algorithm returns the Gibbs free energy of the reactions, the
+concentrations of metabolites and the actual maximum minimum driving force. The optimization
+problem solved is:
+```
+max min -ΔᵣG
+s.t. ΔᵣG = ΔrG⁰ + R T S' ln(C)
+     ΔᵣG ≤ 0
+     ln(Cₗ) ≤ ln(C) ≤ ln(Cᵤ)
+```
+where `ΔrG` are the Gibbs energies dissipated by the reactions, R is the gas constant, T is
+the temperature, S is the stoichiometry of the model, and C is the vector of metabolite
+concentrations (and their respective lower and upper bounds).
+
+In case no feasible solution exists, `nothing` is returned.
+
+Reactions specified in `ignore_reaction_ids` are internally ignored when calculating the
+max-min driving force. This should include water and proton importers.
+
+Since biochemical thermodynamics are assumed, the `proton_ids` and `water_ids` need to be
+specified so that they can be ignored in the calculations. Effectively this assumes an
+aqueous environment at constant pH is used.
+
+`constant_concentrations` is used to fix the concentrations of certain metabolites (such as
+CO₂). `concentration_ratios` is used to specify additional constraints on metabolite pair
+concentrations (typically, this is done with various cofactors such as the ATP/ADP ratio.
+For example, you can fix the concentration of ATP to be always 5× higher than of ADP by
+specifying `Dict(("ATP","ADP") => 5.0)`
+
+`concentration_lb` and `concentration_ub` set the `Cₗ` and `Cᵤ` in the
+optimization problems.
+
+`T` and `R` can be specified in the corresponding units; defaults are K and kJ/K/mol.
+"""
+function max_min_driving_force_analysis(
+    model::A.AbstractFBCModel,
+    reaction_standard_gibbs_free_energies::Dict{String,Float64};
+    flux_solution = Dict{String,Float64}(),
+    concentration_ratios = Dict{String,Tuple{String,String,Float64}}(),
+    proton_ids = ["h_c", "h_e"],
+    water_ids = ["h2o_c", "h2o_e"],
+    concentration_lb = 1e-9,
+    concentration_ub = 100e-3,
```

Comment:
this is unnecessarily confusing and incomparable to the above `1e-9`. It's simply `1e-1`. With `100e-3` people need to go (`-3+2`) to actually see the magnitude.